### PR TITLE
make spid ext. capability

### PR DESCRIPTION
**Pull request by noodlemoodle**
https://github.com/nanoPayinc/NANOPAY/pull/10619

### CODE REVIEW

File: src/foam/nanos/auth/User.js [hunk 1]
```diff
@@ -516,6 +516,22 @@ foam.CLASS({
       name: 'lifecycleState',
       value: foam.nanos.auth.LifecycleState.PENDING,
       writePermissionRequired: true
+    },
+    {
+      class: 'Reference',
+      of: 'foam.nanos.auth.ServiceProvider',
+      name: 'spid',
+      documentation: `
```
Comment (jlhughes) [inline src/foam/nanos/auth/User.js:520]:
```suggestion
      tableWidth: 120,	
      section: 'administrative',	
      documentation: `
```

**Review by jlhughes — APPROVED** [review]

**Comment by lchanmann** [comment]
serviceProviderDAO is authorized with "serviceprovider" as permission prefix - https://github.com/foam-framework/foam2/blob/70455e1aae5d2c43e90a308c39ab8a261b960987/src/services#L592.

@kgrgreer @jlhughes do we want to instead use "capability" as the permission prefix for all filtered DAOs?

File: src/foam/nanos/crunch/PreventMultipleSpidDAO.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
```
Comment (lchanmann) [inline src/foam/nanos/crunch/PreventMultipleSpidDAO.js]:
```suggestion
 * Copyright 2020 The FOAM Authors. All Rights Reserved.
```

File: src/foam/nanos/crunch/PreventMultipleSpidDAO.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  name: 'PreventMultipleSpidDAO',
+  extends: 'foam.dao.ProxyDAO',
+
+  documentation: 'DAO decorator that prevents assigning the user more than one spid',
+
+  javaImports: [
+    'foam.dao.DAO',
+    'foam.mlang.predicate.AbstractPredicate',
+    'foam.nanos.auth.ServiceProvider',
+    'foam.nanos.auth.ServiceProvider',
```
Comment (lchanmann) [inline src/foam/nanos/crunch/PreventMultipleSpidDAO.js]: duplicate import

File: src/foam/nanos/crunch/PreventMultipleSpidDAO.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        ServiceProvider sp = (ServiceProvider) targetCapability;
+
+        AbstractPredicate serviceProviderTargetPredicate = new AbstractPredicate(x) {
+          @Override
+          public boolean f(Object obj) {
+            UserCapabilityJunction ucj = (UserCapabilityJunction) obj;
+            Capability c = (Capability) ucj.findTargetId(x);
+            return c instanceof ServiceProvider;
+          }
+        };
```
Comment (lchanmann) [inline src/foam/nanos/crunch/PreventMultipleSpidDAO.js]: just an idea, if we use a ucj subtype specific for service provider eg. UserServiceProviderJunction then the predicate would just be `INSTANCE_OF(UserServiceProviderJunction.class)` thus avoiding the need to go through association. @jlhughes @kgrgreer what do you think?

**Review by jlhughes — APPROVED** [review]

**Comment by jlhughes** [comment]
@noodlemoodle last request - if you could move all the Spid capabilities - rules, capabilities to the nanos/auth folder, then all MSP stuff is together.

File: src/foam/nanos/crunch/PreventMultipleSpidDAO.js (hunk 4 above)
Comment (jlhughes) [reply src/foam/nanos/crunch/PreventMultipleSpidDAO.js]: @lchanmann I'm not complete familiar with 'association', but the it 'sounds' like the subtype would make the logic cleaner. 
BUT - this should be happening infrequently, as just discussed elsewhere, this is an error condition, so perhaps overkill for the odd use.

**Review by jlhughes — APPROVED** [review]

**Comment by jlhughes** [comment]
> serviceProviderDAO is authorized with "serviceprovider" as permission prefix -
> 
> https://github.com/foam-framework/foam2/blob/70455e1aae5d2c43e90a308c39ab8a261b960987/src/services#L592
> 
> .
> @kgrgreer @jlhughes do we want to instead use "capability" as the permission prefix for all filtered DAOs?

All existing spid permissions use 'spid' as prefix. 
Affects: 
**nanopay/src:** 

- net/nanopay/auth/UserCreateServiceProviderURLRuleAction.js
- net/nanopay/auth/groupPermissionJunctions.jrl
- net/nanopay/msp/MspInfo.js
- permissions
- scripts

**foam/src:**

- scripts
- foam/nanos/auth/test/UserAndGroupPermissionTest.js
- foam/nanos/auth/User.js
- foam/nanos/auth/ServiceProviderAwareDAO.js
- foam/nanos/auth/ServiceProviderAwareSupport.js
- foam/nanos/ruler/Rule.js

So bit of work, but I'm good with moving to 
capability.read.spid. 
question, how does one get something like spid.read.* or that is simply not possible - not that we really need it.

**Comment by jlhughes** [comment]
@lchanmann @noodlemoodle Please re-review. I made the permission prefix changes. All tests pass on nanopy.

**Pull request merged**